Qi: Ambiguity in transform_attribute trait at dealing with Fusion ADT

```cpp
#include <boost/spirit/include/qi.hpp>
#include <boost/spirit/include/support_adapt_adt_attributes.hpp>

struct Vec2i
{
    int x;
    int y;
};

BOOST_FUSION_ADAPT_ADT(
    Vec2i,
    (int, int, obj.x, obj.x = val)
    (int, int, obj.y, obj.y = val)
)

int main()
{
    namespace qi = boost::spirit::qi;

    qi::rule<char const*, Vec2i()> v1 = qi::uint_ >> ',' >> qi::uint_; // OK

    qi::rule<char const*, int()> coord = qi::uint_;
    qi::rule<char const*, Vec2i()> v2 = coord >> ',' >> coord; // Fails to compile
}
```

```
boost\spirit\home\qi\detail\attributes.hpp(154): error C2752: 'boost::spirit::traits::transform_attribute<Exposed,Transformed,boost::spirit::qi::domain,void>': more than one partial specialization matches the template argument list
        with
        [
            Exposed=boost::fusion::extension::adt_attribute_proxy<Vec2i,0,false>,
            Transformed=int
        ]
boost\spirit\home\qi\detail\attributes.hpp(154): note: could be 'boost::spirit::traits::transform_attribute<Exposed,Transformed,boost::spirit::qi::domain,void>'
boost\spirit\home\qi\detail\attributes.hpp(154): note: or       'boost::spirit::traits::transform_attribute<boost::fusion::extension::adt_attribute_proxy<T,N,false>,Attribute,Domain,disable_if<boost::is_reference<fusion::extension::adt_attribute_proxy<T,N,false>::type>,void>::type>'

```

--- a/include/boost/spirit/home/karma/detail/attributes.hpp
+++ b/include/boost/spirit/home/karma/detail/attributes.hpp
@@ -95,14 +95,12 @@ namespace boost { namespace spirit { namespace karma
 }}}
 
 ///////////////////////////////////////////////////////////////////////////////
-namespace boost { namespace spirit { namespace traits
+namespace boost { namespace spirit { namespace traits { namespace detail
 {
     template <typename Exposed, typename Transformed>
-    struct transform_attribute<Exposed, Transformed, karma::domain>
+    struct transform_attribute_base<Exposed, Transformed, karma::domain>
       : karma::transform_attribute<Exposed, Transformed>
     {};
-}}}
+}}}}
 
 #endif
-
-

--- a/include/boost/spirit/home/qi/detail/attributes.hpp
+++ b/include/boost/spirit/home/qi/detail/attributes.hpp
@@ -143,10 +143,12 @@ namespace boost { namespace spirit { namespace qi
 ///////////////////////////////////////////////////////////////////////////////
 namespace boost { namespace spirit { namespace traits
 {
-    template <typename Exposed, typename Transformed>
-    struct transform_attribute<Exposed, Transformed, qi::domain>
-      : qi::transform_attribute<Exposed, Transformed>
-    {};
+    namespace detail {
+        template <typename Exposed, typename Transformed>
+        struct transform_attribute_base<Exposed, Transformed, qi::domain>
+          : qi::transform_attribute<Exposed, Transformed>
+        {};
+    }
 
     template <typename Exposed, typename Transformed>
     struct transform_attribute<Exposed&, Transformed, qi::domain>

--- a/include/boost/spirit/home/support/attributes.hpp
+++ b/include/boost/spirit/home/support/attributes.hpp
@@ -923,6 +923,14 @@ namespace boost { namespace spirit { namespace traits
         type;
     };
 
+    namespace detail {
+        // Domain-agnostic class template partial specializations and
+        // type agnostic domain partial specializations are ambious.
+        // To resolve the ambiguity type agnostic domain partial
+        // specializations are dispatched via intermediate type.
+        template <typename Exposed, typename Transformed, typename Domain>
+        struct transform_attribute_base;
+    }
     ///////////////////////////////////////////////////////////////////////////
     //  transform_attribute
     //
@@ -933,7 +941,9 @@ namespace boost { namespace spirit { namespace traits
     ///////////////////////////////////////////////////////////////////////////
     template <typename Exposed, typename Transformed, typename Domain
       , typename Enable/* = void*/>
-    struct transform_attribute;
+    struct transform_attribute
+      : detail::transform_attribute_base<Exposed, Transformed, Domain>
+    {};
 
     ///////////////////////////////////////////////////////////////////////////
     template <typename Domain, typename Transformed, typename Exposed>

--- a/test/qi/regression_adapt_adt.cpp
+++ b/test/qi/regression_adapt_adt.cpp
@@ -92,5 +92,12 @@ int main()
             data.str == "Hello" && !data.optstr);
     }
 
+    { // GH#396
+        qi::rule<char const*, unsigned()> uint_r = qi::uint_;
+        test1 data;
+        BOOST_TEST(test_attr("123@999", uint_r >> -('@' >> uint_r), data) &&
+            data.var == 123 && data.opt && data.opt.get() == 999);
+    }
+
     return boost::report_errors();
 }

